### PR TITLE
[v7r1 -> v7r2] Add DB version table

### PR DESCRIPTION
**Pull request by TaykYoku**
Adding a DB version table to see if the database structure is outdated and needs to be fixed

BEGINRELEASENOTES

*Core
NEW: DB - add DB version and updateDBVersion method

ENDRELEASENOTES

### CODE REVIEW

File: Core/Base/DB.py [hunk 1]
```diff
@@ -51,5 +75,27 @@ def __init__(self, dbname, fullname, debug=False):
...
+
+        :param str optionName: option name
+        :param defaultValue: default value
+
+        :return: value that inherits the defaultValue type
+    """
     cs_path = getDatabaseSection(self.fullname)
     return gConfig.getValue("/%s/%s" % (cs_path, optionName), defaultValue)
+
+  def updateDBVersion(self, version):
```
Comment (fstagni) [inline Core/Base/DB.py]: Right now this is not called anywhere. Which component should be its client?

Comment (TaykYoku) [reply Core/Base/DB.py]: This will be used in future PR for ProxyDB to change the tables

**Comment by chaen** [comment]
Interesting that you are doing this, since I was considering doing something similar very soon.
Since it is not yet used, and before reinventing the wheel, can you please investigate this ? https://flywaydb.org/

**Comment by chaen** [comment]
There is also this one that I would like to use, probably more and more as we move towards SQLAlchemy for all the systems https://alembic.sqlalchemy.org/en/latest/index.html

**Comment by atsareg** [comment]
The idea here is to include a simple function into any DB class which performs necessary operations to upgrade the underlying DB schema based on the schema version in the class itself as compared with the version recorded in the database. For each version increment an action to update the schema will be coded in. This ensures by definition that the class works with the database schema corresponding to the release version of the class. Also, the schema definition and the migration changes are all kept in the same place under the responsibility of the class owner. The schema updates can be expressed as direct SQL statements or corresponding SQLAlchemy functions if applicable. For such a simple functionality I doubt that we need to create a dependency on so complex third party tools as the ones that you mentioned in the comments above.

**Comment by TaykYoku** [comment]
Used  by PR #4421

**Comment by fstagni** [comment]
IIUC correctly this is used in https://github.com/DIRACGrid/DIRAC/pull/4421/files?utf8=%E2%9C%93&diff=unified&w=1#diff-77e24b88f795c4de190eac19520fece0R215
Am I correct?

I think this PR is an attempt to address a real issue, and I like the initiative, but I have to say that I don't like this proposed solution, for a few simple reasons: 
- it's mixing code with operational changes. 
- those lines of code that perform the DB updates are bound to be used only once. If that's the case, I am more in favor of even  creating (horrible!) update scripts, that don't need further maintenance, and don't even need to be under version control. Which is basically re-stating the same concept of the bullet above: don't mix code with operational changes.
- there's this simple https://github.com/DIRACGrid/management repo for DevOps activity that could be expanded, but as pointed out by Chris there are also third party apps, and not very complicated to use.

**Comment by chaen** [comment]
@atsareg since I had no news, I had a look at the tools myself. The first one is probably not the way to go, because the community version misses some feature I think are important (e.g. dry run). 
The second one looks much more promising.

**Pull request closed without merging**